hack/lib/log/output.sh

```diff
@@ -1,27 +1,52 @@
 #!/bin/bash
 
 # This file contains functions used for writing log messages
-# to stdout and stderr from scripts while they run.
-
-function os::log::with-severity() {
-	local msg=$1
-	local severity=$2
-
-	echo "[$2] ${1}"
-}
-readonly -f os::log::with-severity
+# to stdout and stderr from scripts while they run. `printf`
+# is used in order to support newlines in the output while
+# correctly showing literal escape characters that exist in
+# the mesage being logged. `echo -e` would instead interpret
+# literal escapes in the logged message.
 
+# os::log::info writes the message to stdout
+#
+# Arguments:
+#  - all: message to write
 function os::log::info() {
-	os::log::with-severity "${1}" "INFO"
+	printf '[INFO] %s\n' "$*"
 }
 readonly -f os::log::info
 
+# os::log::warn writes the message to stderr.
+# A warning indicates something went wrong but
+# not so wrong that we cannot recover.
+#
+# Arguments:
+#  - all: message to write
 function os::log::warn() {
-	os::log::with-severity "${1}" "WARNING" 1>&2
+	os::text::print_yellow "$( printf '[WARNING] %s\n' "$*" )" 1>&2
```

Comment:
Why use `printf` here instead of string interpolation, like the following?
```bash
os::text::print_yellow "[WARNING] $*" 1>&2;
```
One difference between using string interpolation and using `printf` in a command substitution is that command substitution will chomp all trailing whitespace:
```console
$ function os::log::warn() {
>   os::text::print_yellow "$( printf '[WARNING] %s\n' "$*" )" 1>&2
> }
$ function os::log::warn2() {
>   os::text::print_yellow "[WARNING] $*" 1>&2;
> }
$ os::log::warn $'foo\n\n\n'
[WARNING] foo
$ os::log::warn2 $'foo\n\n\n'
[WARNING] foo



$
```
Was that the intention?